### PR TITLE
HDDS-10573. [hsync] Add a freon tool to benchmark hsync/write concurrency

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

A simple Freon tool to generate concurrent write and hsync workload to one or multiple files.

Sample usage and output:
```
ozone freon hsync-generator -t 20 --bytes-per-write=1024 --path=ofs://ozone1/vol1/bucket1/p -n 500000 --number-of-files=2

             count = 500000
         mean rate = 2868.67 calls/second
     1-minute rate = 2969.64 calls/second
     5-minute rate = 2734.69 calls/second
    15-minute rate = 2644.31 calls/second
               min = 1.63 milliseconds
               max = 145.83 milliseconds
              mean = 6.21 milliseconds
            stddev = 8.77 milliseconds
            median = 4.07 milliseconds
              75% <= 6.22 milliseconds
              95% <= 12.60 milliseconds
              98% <= 37.96 milliseconds
              99% <= 53.12 milliseconds
            99.9% <= 111.50 milliseconds
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10573

## How was this patch tested?

Unit test.

### CODE REVIEW

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+  @Override
+  public Void call() throws Exception {
+    init();
+
+    if (configuration == null) {
+      configuration = freon.createOzoneConfiguration();
+    }
+    URI uri = URI.create(rootPath);
+    fileSystem = FileSystem.get(uri, configuration);
```
Comment (jojochuang) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:100]: the fileSystem object should be closed at the end of the call too.

I'm seeing this message when the tool stops which maybe related.

24/07/11 23:40:37 WARN grpc.GrpcUtil: Timed out gracefully shutting down connection: ManagedChannelOrphanWrapper{delegate=ManagedChannelImpl{logId=7, target=10.140.22.129:9858}}.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestHsyncGenerator.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    String volumeName = "vol1";
+    String bucketName = "bucket1";
+    try (OzoneClient client = cluster.newClient()) {
+      ObjectStore store = client.getObjectStore();
+      store.createVolume(volumeName);
+      OzoneVolume volume = store.getVolume(volumeName);
+      volume.createBucket(bucketName);
+
+      //String rootPath = "o3fs://" + bucketName + "." + volumeName;
```
Comment (jojochuang) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestHsyncGenerator.java:97]: remove

**Review by jojochuang — COMMENTED** [review]
Mostly looks good just a few small nits

**Review by jojochuang — APPROVED** [review]
+1 thanks @duongkame

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (ivandika3) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:100]: Do you think we should incorporate thread-local instance of `FileSystem` as done in https://github.com/apache/ozone/pull/1479.

Note: I'm currently working to unify freon tests in https://issues.apache.org/jira/browse/HDDS-11081, hopefully a patch will be raised soon.

**Review by ivandika3 — COMMENTED** [review]
@duongkame Thanks for the patch. Overall looks good. Left one comment.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java (hunk 3 above)
Comment (duongkame) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:100]: @duongkame We unlikely need more than one FileSystem in this tool. In https://github.com/apache/ozone/pull/1479, the  `HadoopFsGenerator` creates many files, (thus a big load to OM to create file metadata). 
This Freon tool creates one or a few files upfront.

Comment (duongkame) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java]: Anyway, I separated FileSystem per outputstream.

**Review by smengcl — APPROVED** [review]

**Comment by jojochuang** [comment]
Merged. Thanks @smengcl @duongkame and @ivandika3

**Pull request merged**